File: python/linked-list/linked_list.py
class Node(object):
    def __init__(self, value, succeeding=None, previous=None):
        self.value = value
        self.succeeding = succeeding
        self.previous = previous


class LinkedList(object):

    def __init__(self):
        self.head = None
        self.tail = None

    def push(self, value):
        new_node = Node(value)
        if self.tail is not None:
            self.tail.succeeding = new_node
            new_node.previous = self.tail
        self.tail = new_node
        if self.head is None:
            self.head = new_node

    def pop(self):
        if self.tail is not None:
            popped_node = self.tail
            self.tail = popped_node.previous
            return popped_node.value

    def unshift(self, value):
        new_node = Node(value)
        if self.head is not None:
            self.head.previous = new_node
            new_node.succeeding = self.head
        self.head = new_node
        if self.tail is None:
            self.tail = new_node

    def shift(self):
        if self.head is not None:
            shifted_node = self.head
            self.head = shifted_node.succeeding
            return shifted_node.value

File: typescript/linked-list/linked-list.ts
class Node<T> {
    public next: Node<T> | null;
    public data: T;

    constructor(data: T, next?: Node<T>) {
        this.data = data;
        this.next = next ?? null;
    }
}
export default class LinkedList<T> {
    private head: Node<T> | null;

    constructor() {
        this.head = null;
    }

    public push(data: T): void {
        const newNode = new Node(data)

        if (this.head == null) {
            this.head = newNode;
        } else {
            const lastNode = this.getLastNode(this.head);
            lastNode.next = newNode;
        }
    }
    public pop(): T {
        if (this.head == null) {
            throw new Error("Cannot call pop on an empty linked list")
        } else if (this.head.next == null) {
            const node = this.head;
            this.head = null;
            return node.data;
        } else {
            const secondToLastNode = this.getSecondToLastNode(this.head);
            const lastNode = secondToLastNode.next;

            if (lastNode == null) {
                // We don't expect this to happen
                throw new Error("Last node is null")
            } else {
                secondToLastNode.next = null;
                return lastNode.data;
            }
        }
    }
    public shift(): T {
        if (this.head == null) {
            throw new Error("Cannot call shift on an empty linked list")
        } else if (this.head.next == null) {
            const node = this.head;
            this.head = null;
            return node.data;
        } else {
            const node = this.head;
            this.head = node.next;
            return node.data;
        }
    }
    public unshift(data: T): void {
        const newNode = new Node(data)

        if (this.head == null) {
            this.head = newNode;
        } else {
            newNode.next = this.head
            this.head = newNode
        }
    }
    public delete(data: T): void {
        console.log(`Deleting ${data}`);
    }
    public count(): number {
        return 0;
    }

    private getLastNode(node: Node<T>): Node<T> {
        let current = node;
        while (current.next != null) {
            current = current.next;
        }
        return current;
    }

    private getSecondToLastNode(node: Node<T>): Node<T> {
        let current = node;
        while (current.next != null && current.next.next != null) {
            current = current.next;
        }
        return current;
    }
}
